From a96a4bbdfbc402669d6829c549723bfe800774dc Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Tue, 11 Aug 2026 09:04:18 +0000
Subject: [PATCH] Detect Moto dynamic hardware feature

Moto added a custom node in sysconfig XMLs:
<unavailable-feature-conditional />
This node reads a property and enables a feature based on it.

Take those into account to enable NFC on Moto devices which have
NFC-less variants

[Port to A16] Switch refactored to helper-method style; new case inserted inline between unavailable-feature and allow-in-power-save-except-idle with identical logic.

Change-Id: I94ddfbf857a8c55ab67e504f76cc1acf2028e437
---
 .../java/com/android/server/SystemConfig.java | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/services/core/java/com/android/server/SystemConfig.java b/services/core/java/com/android/server/SystemConfig.java
index 2dca73ff7864..731a418cf6df 100644
--- a/services/core/java/com/android/server/SystemConfig.java
+++ b/services/core/java/com/android/server/SystemConfig.java
@@ -926,6 +926,25 @@ public class SystemConfig {
                     case "unavailable-feature":
                         readUnavailableFeature(parser, permFile, allowFeatures, name);
                         break;
+                    case "unavailable-feature-conditional": {
+                        if (allowFeatures) {
+                            String fname = parser.getAttributeValue(null, "name");
+                            String prop = parser.getAttributeValue(null, "prop");
+                            if (fname == null || prop == null) {
+                                Slog.w(TAG, "<" + name + "> without name in " + permFile
+                                        + " at " + parser.getPositionDescription());
+                            } else {
+                                if(android.os.SystemProperties.getBoolean(prop, false)) {
+                                    addFeature(fname, 0);
+                                } else {
+                                    mUnavailableFeatures.add(fname);
+                                }
+                            }
+                        } else {
+                            logNotAllowedInPartition(name, permFile, parser);
+                        }
+                        XmlUtils.skipCurrentTag(parser);
+                    } break;
                     case "allow-in-power-save-except-idle":
                         readAllowInPowerSaveExceptIdle(
                                 parser, permFile, allowOverrideAppRestrictions, name);
-- 
2.43.0

